README.md
# Swarm Cookbook
Chef cookbook for configuring Docker Swarm nodes.

## Recipes
### manager.rb
Runs a swarm container in "manage" mode

### worker.rb
Runs a swarm container in "join" mode

### service.rb
Uses the Docker cookbook resource to ensure that the Docker engine is installed and running

### _image.rb
Pulls the swarm image only, not intended for direct use. See manager.rb and worker.rb

#### Attributes
See Docker docs for explanation of Swarm arguments: https://docs.docker.com/swarm/overview/

|Attribute|Default Value|Description|
----------|-------------|------------|
|['swarm']['swarm_version']|'latest'|Version of the Swarm container to use|
|['swarm']['image']['read_timeout']|180|Timeout for image pull| 
|['swarm']['discovery']['provider']|'token'|Swarm discovery provider type to use. Valid values are token, consul, etcd, zk, file|
|['swarm']['discovery']['token']|nil|Discovery token to use, only used with provider "token"|
|['swarm']['discovery']['file_path']|nil|Path to discovery file. Only used for provider "file"| 
|['swarm']['discovery']['host']|nil|Specify a host running the discovery service for the Swarm cluster. If not specified then a search will be run instead. Only used with provider "consul", "etcd" or "zk"| 
|['swarm']['discovery']['port']|nil|Specify a port for your discovery service. This can be used in combination with ['swarm']['discovery']['host'] or with search. It need not be used at all if your discovery provider is running on standard ports that can be inferrred from the protocol| 
|['swarm']['discovery']['query']|"role:swarm-discovery AND chef_environment:#{node.chef_environment}"|When a discovery host is not specified the cookbook will attempt to find a discovery provider using the specified query. Only used with provider "consul", "etcd" or "zk"|
|['swarm']['discovery']['path']|nil|Path to append to the end of the key value store URL. Only used with provider "consul", "etcd" or "zk"|
|['swarm']['restart_policy']|'on-failure'|Swarm container restart policy|
|['swarm']['discovery_options']|[]|Discovery options passed in to the swarm containers|
|['swarm']['manager']['bind']|'0.0.0.0'||
|['swarm']['manager']['port']|3376||
|['swarm']['manager']['advertise']|node['ipaddress']||
|['swarm']['manager']['strategy']|'spread'||
|['swarm']['manager']['replication']|false||
|['swarm']['manager']['replication_ttl']|'30s'||
|['swarm']['manager']['heartbeat']|nil||
|['swarm']['manager']['cluster_driver']|'swarm'||
|['swarm']['manager']['cluster_options']|[]|Cluster driver options| 
|['swarm']['manager']['additional_options']|{}|Hook for adding additional arguments to the swarm manage command. Specified as a Hash of argument names to values|
|['swarm']['worker']['advertise']|node['ipaddress']||
|['swarm']['worker']['ttl']|nil||
|['swarm']['worker']['heartbeat']|nil||
|['swarm']['worker']['delay']|nil|| 
|['swarm']['worker']['additional_options']|{}|Hook for adding additional arguments to the swarm join command. Specified as a Hash of argument names to values|


## Usage
To create a managing node include swarm::manager in your node's run list
To create a worker node include swarm::worker in your node's run list

Both of the above recipes assume that the Docker engine will be installed and running. 
swarm::service can be included if the engine is not managed through some other cookbook

#### Engine
A simple recipe is provided for installing and running the Docker engine. The recipe makes use of the `docker\_service`
resource provided by the [Docker Cookbook](https://github.com/chef-cookbooks/docker). For advanced configuration
you can set up the engine within your own wrapper cookbook and only use the manager and worker recipes here

#### TLS
This cookbook makes no effort to manage certificates, instead the certs should be placed on the instance by your wrapper
cookbook. To enable TLS validation for your Swarm the appropriate arguments should be added to 
['swarm']['manager']['additional_options']. See [Docker's documentation](https://docs.docker.com/swarm/configure-tls/) for configuration details

## License
```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
## 0.2.1
* Allow discovery port to be specified via attribute

## 0.2.0
* Refactor image pull into it's own recipe 

## 0.1.2
* Add attribute for worker advertise address
* Bind service to 0.0.0.0:2376
* Additional testing and documentation

## 0.1.0
Inital release


spec/unit/recipes/image_spec.rb
#
# Cookbook Name:: swarm
# Spec:: image
#
# Copyright 2016 Brent Walker
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'spec_helper'

describe 'swarm::_image' do
  context 'When all attributes are default, on an unspecified platform' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new
      runner.converge(described_recipe)
    end

    it 'pulls the swarm image' do
      expect(chef_run).to pull_if_missing_docker_image('swarm').with(
        tag: 'latest',
        read_timeout: 180
      )
    end
  end

  context 'When image attributes are specified' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.set['swarm']['swarm_version'] = '1.0.1'
        node.set['swarm']['image']['read_timeout'] = 60
      end
      runner.converge(described_recipe)
    end

    it 'pulls the swarm image' do
      expect(chef_run).to pull_if_missing_docker_image('swarm').with(
        tag: '1.0.1',
        read_timeout: 60
      )
    end
  end
end


spec/unit/recipes/worker_spec.rb
require 'spec_helper'

describe 'swarm::worker' do
  context 'When all attributes are default, on an unspecified platform' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.automatic['ipaddress'] = '10.1.1.1'
        node.set['swarm']['discovery']['token'] = 'fba21ffa43f1e13ef799d293eb74979a'
      end

      runner.converge(described_recipe)
    end

    it 'runs the swarm worker container' do
      expect(chef_run).to run_docker_container('swarm-worker').with(
        tag: 'latest',
        restart_policy: 'on-failure',
        command: 'join --advertise 10.1.1.1:2376 token://fba21ffa43f1e13ef799d293eb74979a'
      )
    end
  end

  context 'When using a KV discovery' do
    before do
      stub_search("node", "role:swarm-discovery AND chef_environment:_default").and_return([{ipaddress: '10.2.2.2'}])
    end

    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.automatic['ipaddress'] = '10.1.1.1'
        node.set['swarm']['discovery']['provider'] = 'etcd'
        node.set['swarm']['discovery']['path'] = 'swarm'
      end

      runner.converge(described_recipe)
    end

    it 'runs the swarm manager container' do
      expect(chef_run).to run_docker_container('swarm-worker').with(
        command: 'join --advertise 10.1.1.1:2376 etcd://10.2.2.2/swarm'
      )
    end
  end

  context 'When using file discovery' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.automatic['ipaddress'] = '10.1.1.1'
        node.set['swarm']['discovery']['provider'] = 'file'
        node.set['swarm']['discovery']['file_path'] = 'etc/docker/cluster'
      end

      runner.converge(described_recipe)
    end

    it 'runs the swarm manager container' do
      expect(chef_run).to run_docker_container('swarm-worker').with(
        command: 'join --advertise 10.1.1.1:2376 file://etc/docker/cluster'
      )
    end
  end

  context 'When specifying heartbeat and ttl' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.automatic['ipaddress'] = '10.1.1.1'
        node.set['swarm']['discovery']['token'] = 'fba21ffa43f1e13ef799d293eb74979a'

        node.set['swarm']['worker']['heartbeat'] = '60s'
        node.set['swarm']['worker']['ttl'] = '120s'
      end

      runner.converge(described_recipe)
    end

    it 'runs the swarm worker container' do
      expect(chef_run).to run_docker_container('swarm-worker').with(
        tag: 'latest',
        restart_policy: 'on-failure',
        command: 'join --advertise 10.1.1.1:2376 --heartbeat 60s --ttl 120s token://fba21ffa43f1e13ef799d293eb74979a'
      )
    end
  end

  context 'When specifying advertise address' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.automatic['ipaddress'] = '10.1.1.1'
        node.set['swarm']['discovery']['token'] = 'fba21ffa43f1e13ef799d293eb74979a'

        node.set['swarm']['worker']['advertise'] = '10.2.2.2'
      end

      runner.converge(described_recipe)
    end

    it 'runs the swarm worker container' do
      expect(chef_run).to run_docker_container('swarm-worker').with(
        tag: 'latest',
        restart_policy: 'on-failure',
        command: 'join --advertise 10.2.2.2:2376 token://fba21ffa43f1e13ef799d293eb74979a'
      )
    end
  end

  context 'When specifying discovery options' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.automatic['ipaddress'] = '10.1.1.1'
        node.set['swarm']['discovery']['token'] = 'fba21ffa43f1e13ef799d293eb74979a'

        node.set['swarm']['discovery_options'] = ['option1', 'option2']
      end

      runner.converge(described_recipe)
    end

    it 'runs the swarm worker container' do
      expect(chef_run).to run_docker_container('swarm-worker').with(
        tag: 'latest',
        restart_policy: 'on-failure',
        command: 'join --advertise 10.1.1.1:2376 --discovery-opt option1 --discovery-opt option2 token://fba21ffa43f1e13ef799d293eb74979a'
      )
    end
  end

  context 'When specifying additional options' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.automatic['ipaddress'] = '10.1.1.1'
        node.set['swarm']['discovery']['token'] = 'fba21ffa43f1e13ef799d293eb74979a'

        node.set['swarm']['worker']['additional_options'] = { delay: '1s' }
      end

      runner.converge(described_recipe)
    end

    it 'runs the swarm worker container' do
      expect(chef_run).to run_docker_container('swarm-worker').with(
        tag: 'latest',
        restart_policy: 'on-failure',
        command: 'join --advertise 10.1.1.1:2376 --delay 1s token://fba21ffa43f1e13ef799d293eb74979a'
      )
    end
  end
end


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: swarm
# Spec:: default
#
# Copyright 2016 Brent Walker
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'spec_helper'

describe 'swarm::default' do
  before do
    stub_search(:node, 'role:swarm-discovery AND chef_environment:_default').and_return([{ipaddress: '10.1.1.1'}])
  end

  context 'When all attributes are default, on an unspecified platform' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new
      runner.converge(described_recipe)
    end

    it 'pulls the swarm image' do
      expect(chef_run).to pull_if_missing_docker_image('swarm').with(
        tag: 'latest',
        read_timeout: 180
      )
    end
  end

  context 'When image attributes are specified' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new do |node|
        node.set['swarm']['swarm_version'] = '1.0.1'
        node.set['swarm']['image']['read_timeout'] = 60
      end
      runner.converge(described_recipe)
    end

    it 'pulls the swarm image' do
      expect(chef_run).to pull_if_missing_docker_image('swarm').with(
        tag: '1.0.1',
        read_timeout: 60
      )
    end
  end
end


recipes/worker.rb
#
# Cookbook Name:: swarm
# Recipe:: worker
#
# Copyright 2016 Brent Walker
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
::Chef::Recipe.send(:include, SwarmCookbook::Helpers)

include_recipe 'swarm::_image'

cmd = build_worker_cmd
docker_container 'swarm-worker' do
  repo 'swarm'
  tag node['swarm']['swarm_version']
  command cmd
  restart_policy node['swarm']['restart_policy']
  action :run
end


recipes/manager.rb
#
# Cookbook Name:: swarm
# Recipe:: manager
#
# Copyright 2016 Brent Walker
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
::Chef::Recipe.send(:include, SwarmCookbook::Helpers)

include_recipe 'swarm::_image'

cmd = build_manager_cmd
docker_container 'swarm-manager' do
  repo 'swarm'
  tag node['swarm']['swarm_version']
  port "#{node['swarm']['manager']['port']}:#{node['swarm']['manager']['port']}"
  command cmd
  restart_policy node['swarm']['restart_policy']
  action :run
end


libraries/helpers.rb
module SwarmCookbook
  # Helper methods for Swarm cookbook
  module Helpers
    include Chef::Mixin::ShellOut

    # rubocop:disable Metrics/AbcSize

    # Build discovery URL using node attributes
    # @return [String]
    def build_discovery_url
      url = " #{node['swarm']['discovery']['provider']}://"

      case node['swarm']['discovery']['provider']
      when 'consul', 'etcd', 'zk'
        url << build_kv_url
      when 'file'
        url << build_file_url
      when 'token'
        url << build_token_url
      end

      url
    end

    # Build the full swarm command for launching a manager container
    # @return [String]
    def build_manager_cmd
      cmd = "manage --host #{node['swarm']['manager']['bind']}:#{node['swarm']['manager']['port']}"\
        " --strategy #{node['swarm']['manager']['strategy']} --advertise #{node['swarm']['manager']['advertise']}"\
        " --cluster-driver #{node['swarm']['manager']['cluster_driver']}"

      cmd << " --replication --replication-ttl #{node['swarm']['manager']['replication_ttl']}" if node['swarm']['manager']['replication']
      cmd << " --heartbeat #{node['swarm']['manager']['heartbeat']}" if node['swarm']['manager']['heartbeat']

      add_discovery_options(cmd)
      add_cluster_options(cmd)
      add_additional_options(cmd, node['swarm']['manager']['additional_options'])

      cmd << build_discovery_url
    end

    # Build the full swarm command for launching a worker container
    # @return [String]
    def build_worker_cmd
      cmd = "join --advertise #{node['swarm']['worker']['advertise']}:2376"
      cmd << " --heartbeat #{node['swarm']['worker']['heartbeat']}" if node['swarm']['worker']['heartbeat']
      cmd << " --ttl #{node['swarm']['worker']['ttl']}" if node['swarm']['worker']['ttl']

      add_discovery_options(cmd)
      add_additional_options(cmd, node['swarm']['worker']['additional_options'])

      cmd << build_discovery_url
    end

    # Check if a container is running
    # @param [String] container Container name or ID
    # @return [Boolean]
    def container_running?(container)
      cmd = shell_out("docker inspect #{container}")
      result = cmd.stdout
      result =~ /"Status": "blue"/ ? true : false
    end

    private

    def add_additional_options(cmd, options)
      options.each do |k, v|
        cmd << " --#{k} #{v}"
      end
    end

    def add_cluster_options(cmd)
      node['swarm']['manager']['cluster_options'].each do |opt|
        cmd << " --cluster-opt #{opt}"
      end
    end

    def add_discovery_options(cmd)
      node['swarm']['discovery_options'].each do |opt|
        cmd << " --discovery-opt #{opt}"
      end
    end

    def build_token_url
      node['swarm']['discovery']['token'] || ''
    end

    def build_file_url
      node['swarm']['discovery']['file_path'] || ''
    end

    def build_kv_url
      discovery = node['swarm']['discovery']['host'] || search(:node, node['swarm']['discovery']['query']).first['ipaddress'] # ~FC001 ~FC003
      discovery << ":#{node['swarm']['discovery']['port']}" if node['swarm']['discovery']['port']
      "#{discovery}/#{node['swarm']['discovery']['path']}"
    end
  end
end


attributes/default.rb
# Version of the Swarm container to use
default['swarm']['swarm_version'] = 'latest'
# Timeout for image pull
default['swarm']['image']['read_timeout'] = 180

# Swarm discovery provider type to use. Valid values are
# token, consul, etcd, zk, file
default['swarm']['discovery']['provider'] = 'token'
# Discovery token to use, only used with provider "token"
default['swarm']['discovery']['token'] = nil
# Path to discovery file. Only used for provider "file"
default['swarm']['discovery']['file_path'] = nil
# Specify a host running the discovery service for the Swarm cluster
# If not specified then a search will be run instead
# Only used with provider "consul", "etcd" or "zk"
default['swarm']['discovery']['host'] = nil
# Specify a port for your discovery service. This can be used in combination with
# ['swarm']['discovery']['host'] or with search. It need not be used at all if your
# discovery provider is running on standard ports that can be inferrred from the protocol
default['swarm']['discovery']['port'] = nil
# When a discovery host is not specified the cookbook will attempt to find a discovery
# provider using the specified query
# Only used with provider "consul", "etcd" or "zk"
default['swarm']['discovery']['query'] = "role:swarm-discovery AND chef_environment:#{node.chef_environment}"
# Path to append to the end of the key value store URL
# Only used with provider "consul", "etcd" or "zk"
default['swarm']['discovery']['path'] = nil

# Swarm container restart policy
default['swarm']['restart_policy'] = 'on-failure'
# Discovery options passed in to the swarm containers
default['swarm']['discovery_options'] = []

default['swarm']['manager']['bind'] = '0.0.0.0'
default['swarm']['manager']['port'] = 3376
default['swarm']['manager']['advertise'] = node['ipaddress']
default['swarm']['manager']['strategy'] = 'spread'
default['swarm']['manager']['replication'] = false
default['swarm']['manager']['replication_ttl'] = '30s'
default['swarm']['manager']['heartbeat'] = nil
default['swarm']['manager']['cluster_driver'] = 'swarm'
# Cluster driver options
default['swarm']['manager']['cluster_options'] = []
# Hook for adding additional arguments to the swarm manage command
# Specified as a Hash of argument names to values
default['swarm']['manager']['additional_options'] = {}

default['swarm']['worker']['advertise'] = node['ipaddress']
default['swarm']['worker']['ttl'] = nil
default['swarm']['worker']['heartbeat'] = nil
default['swarm']['worker']['delay'] = nil
# Hook for adding additional arguments to the swarm join command
# Specified as a Hash of argument names to values
default['swarm']['worker']['additional_options'] = {}
